### PR TITLE
[test] Use test-utils from npm

### PR DESCRIPTION
**Pull request by michaldudak**
Using test-utils from an npm package instead of monorepo.

A couple of additional changes were needed:
- tsconfig's `moduleResolution` was set to `bundler` to enable reading package.json `exports` field
- @testing-library/react was upgraded to 15.0.7 to match the version from the test-utils
- @types/react-dom were upgraded to 18.2.25 (the previous version did not contain a type used by testing-library)

To be merged after https://github.com/mui/material-ui/pull/41177 is merged and published.

### CODE REVIEW

File: pnpm-lock.yaml [hunk 1]
```diff
@@ -11544,28 +11448,63 @@ packages:
...
-      ws: 8.16.0
+      ws: 8.17.0
       xml-name-validator: 5.0.0
     transitivePeerDependencies:
       - bufferutil
       - supports-color
       - utf-8-validate
     dev: true
 
+  /jsdom@24.1.0:
```
Comment (LukasTy) [inline pnpm-lock.yaml:11473]: Could we bump `jsdom` to avoid the duplicate dependency? 🤔

File: pnpm-lock.yaml [hunk 2]
```diff
@@ -17438,6 +17325,47 @@ packages:
...
+      fs-extra: 11.2.0
+      jsdom: 24.1.0
+      lodash: 4.17.21
+      mocha: 10.4.0
+      playwright: 1.44.1
+      prop-types: 15.8.1
+      react: 18.2.0
+      react-dom: 18.2.0(react@18.2.0)
+      react-test-renderer: 18.2.0(react@18.2.0)
+      sinon: 15.2.0
```
Comment (LukasTy) [inline pnpm-lock.yaml:17360]: Nitpick: mui-x is on `sinon@v16` already, have you considered bumping it to on monorepo to have it in sync or is there a roadblock for it? 🤔

**Review by LukasTy — COMMENTED** [review]
Nice! 👏

File: pnpm-lock.yaml (hunk 1 above)
Comment (michaldudak) [reply pnpm-lock.yaml:11473]: Sure!

File: pnpm-lock.yaml (hunk 2 above)
Comment (michaldudak) [reply pnpm-lock.yaml:17360]: All good here. It turns out it was jsdom's fault.

**Comment by LukasTy** [comment]
It's clear that the explicit memory limit increase is needed as before it the CPU is pegged doing too much GC. 🙈 

### Before
Screenshot 2024-05-29 at 09 03 34

### After
Screenshot 2024-05-29 at 09 05 37

**Comment by michaldudak** [comment]
https://github.com/mui/material-ui/pull/41177 has been merged and @mui/internal-test-utils is published to npm with version 1.0.0-dev.20240529-082515-213b5e33ab

File: renovate.json [hunk 3]
```diff
@@ -6,7 +6,7 @@
   "commitMessageTopic": "{{depName}}",
   "dependencyDashboard": true,
   "rebaseWhen": "conflicted",
-  "ignoreDeps": ["jsdom"],
+  "ignoreDeps": [],
```
Comment (LukasTy) [inline renovate.json:9]: @cherniavskii On the latest major they reverted a change that initially reduced the performance
https://github.com/jsdom/jsdom/releases/tag/24.0.0

**Review by LukasTy — APPROVED** [review]
LGTM! 👏 
Thanks for the effort. 💙 

I've also updated `jsdom` and ran `dedupe` to reduce amount of duplicate packages, which were introduced for some reason. 🙈 

Do you have any idea why argos reacted like there is no reference build and treats all screenshots as new? 🤔 🤷

**Comment by michaldudak** [comment]
> Do you have any idea why argos reacted like there is no reference build and treats all screenshots as new? 🤔 🤷

It looks like an issue on Argos' side. I started seeing it in other PRs as well.

File: package.json [hunk 4]
```diff
@@ -32,7 +32,7 @@
     "size:why": "pnpm size:snapshot --analyze --accurateBundles",
     "tc": "node test/cli.js",
     "test": "node scripts/test.mjs",
-    "test:coverage": "cross-env NODE_ENV=test TZ=UTC BABEL_ENV=coverage nyc mocha --exclude '**/node_modules/**' && nyc report -r lcovonly",
+    "test:coverage": "cross-env NODE_OPTIONS=--max-old-space-size=4096 NODE_ENV=test TZ=UTC BABEL_ENV=coverage nyc mocha --exclude '**/node_modules/**' && nyc report -r lcovonly",
```
Comment (cherniavskii) [inline package.json:35]: @LukasTy How did you manage to make the `test_unit` pass? It failed even when I added this flag here

Comment (LukasTy) [reply package.json:35]: Did it? I saw them passing even before bumping jsdom... 🤷 
https://app.circleci.com/pipelines/github/mui/mui-x?branch=pull%2F12880

Comment (cherniavskii) [reply package.json:35]: Ahh, I think I didn't wait for the result after my last commit, and for some reason assumed it failed 😅
All good then 👍🏻

Comment (LukasTy) [reply package.json:35]: P.S. Here is my comment comparing your commits behavior on the system resources. 😉 
https://github.com/mui/mui-x/pull/12880#issuecomment-2136584644

File: packages/x-data-grid/package.json [hunk 5]
```diff
@@ -48,6 +48,7 @@
   },
   "dependencies": {
     "@babel/runtime": "^7.24.5",
+    "@mui/internal-test-utils": "1.0.0-dev.20240529-082515-213b5e33ab",
```
Comment (AviVahl) [inline packages/x-data-grid/package.json:51]: @LukasTy this was added as a dependency, causing consumers to get this, and many others deps, installed.
this addition cause peerDep errors (about babel missing and react@18 not supported) on our end.

Comment (LukasTy) [reply packages/x-data-grid/package.json:51]: Thank you for raising this! 🙏 
An issue has also been created for it: https://github.com/mui/mui-x/issues/13317.

**Pull request merged**